File: python/src/aoc/day1.py
from itertools import combinations
from math import prod


def solve_1(data):
    return solve(data, tuple_size=2)


def solve_2(data):
    return solve(data, tuple_size=3)


def solve(numbers, tuple_size):
    for candidate in combinations(numbers, tuple_size):
        if sum(candidate) == 2020:
            return prod(candidate)

File: clojure/src/aoc/day1.clj
(ns aoc.day1
  (:require [clojure.math.combinatorics :refer [combinations]]
            [clojure.test :refer [deftest testing is run-tests]]))

;; Solution

(defn- sum-equals? [number]
  (fn [numbers] (= number (apply + numbers))))

(defn solve [data tuple-size]
  (->> (combinations data tuple-size)
       (filter (sum-equals? 2020))
       first
       (apply *)))


;; Verification

(deftest day-1-verification
  (let [example-data [1721 979 366 299 675 1456]]
    (testing "Part One"
      (is (= (solve example-data 2) 514579)))
    (testing "Part Two"
      (is (= (solve example-data 3) 241861950)))))

(run-tests)
